Clockwork/Support/Vanilla/ClockworkMiddleware.php

```diff
@@ -0,0 +1,80 @@
+<?php namespace Clockwork\Support\Vanilla;
+
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+
+// Generic Clockwork middleware fro use with PSR-compatible applications
+class ClockworkMiddleware implements MiddlewareInterface
+{
+	// The underlying vanilla Clockwork support class instance
+	protected $clockwork;
+
+	// Whether this middleware should handle routing for Clockwork REST api and Web interface
+	protected $handleRouting = true;
+	
+	// PSR-17 request factory instance
+	protected $responseFactory;
+	
+	// Creates new middleware instance, takes a vanilla Clockwork support class instance as an argument
+	public function __construct(Clockwork $clockwork)
+	{
+		$this->clockwork = $clockwork;
+	}
+	
+	// Returns a new middleware instance with a default singleton Clockwork instance, takes an additional configuration as argument
+	public static function init($config = [])
+	{
+		return new static(Clockwork::init($config));
+	}
+	
+	// Sets a PSR-17 compatible response factory. When using the middleware with routing enabled, response factory must be manually set
+	// or the php-http/discovery has to be intalled for zero-configuration use
+	public function withResponseFactory(ResponseFactoryInterface $responseFactory)
+	{
+		$this->responseFactory = $responseFactory;
+		return $this;
+	}
+	
+	// Disables routing handling in the middleware. When disabled additional manual configuration of the application router is required.
+	public function withoutRouting()
+	{
+		$this->handleRouting = false;
+		return $this;
+	}
+	
+	// Process the middleware
+	public function process(ServerRequestInterface $request, RequestHandlerInterface $handler) :ResponseInterface
+	{
+		$request->withAttribute('clockwork', $this->clockwork);
```

Comment:
PSR requests are immutable, so this doesn't do what you intend. It returns a new request with the attribute attached.